pkg/flare/service_windows.go

```diff
@@ -0,0 +1,286 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build windows
+
+package flare
+
+import (
+	"fmt"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+const (
+	scManagerAccess = windows.SC_MANAGER_CONNECT | windows.SC_MANAGER_ENUMERATE_SERVICE
+)
+
+type serviceInfo struct {
+	ServiceName           string
+	ServiceState          string
+	ProcessID             uint32
+	Config                queryServiceConfig
+	DependentServices     []string
+	TriggersCount         uint32
+	ServiceFailureActions serviceFailureActions
+}
+
+type queryServiceConfig struct {
+	ServiceType      string
+	StartType        string
+	ErrorControl     string
+	BinaryPathName   string // fully qualified path to the service binary file, can also include arguments for an auto-start service
+	LoadOrderGroup   string
+	TagID            uint32
+	Dependencies     []string
+	ServiceStartName string // name of the account under which the service should run
+	DisplayName      string
+	Password         string
+	Description      string
+	SidType          uint32 // one of SERVICE_SID_TYPE, the type of sid to use for the service
+	DelayedAutoStart bool   // the service is started after other auto-start services are started plus a short delay
+}
+
+type recoveryActionsUpdated struct {
+	Type  string
+	Delay time.Duration
+}
+
+type serviceFailureActions struct {
+	ResetPeriod                      uint32
+	RebootMessage                    string
+	RecoveryCommand                  string
+	FailureActionsOnNonCrashFailures bool
+	RecoveryActions                  []recoveryActionsUpdated
+}
+
+func getFailureActions(s *mgr.Service) (serviceFailureActions, error) {
+	sfa := serviceFailureActions{}
+
+	var err error
+	var err1 error
+
+	sfa.ResetPeriod, err1 = s.ResetPeriod()
+	if err1 != nil {
+		err = fmt.Errorf("%v, ResetPeriod: %v", err, err1)
```

Comment:
Why not return early in case of errors?

Reply:
The idea here would be to continue with trying to attempt to populate the other fields in the struct, `serviceFailureActions`. Since this entire code is composed of a struct with multiple fields that are collected through different Windows API calls, we don't want the function to stop collecting the other information due an error collecting a single field.

Reply:
It's also why each time an error is returned we're appending it to `err` to which is returned at the end of the function.

Reply:
It is cleave approach. I like it. I do not think it worth to change, However it is opportunity to express my $0.02. Intuitively at least, and I could be wrong. It is unlikely that one of the API fails and other succeed. But immediate return per @sgnn7 makes code much simpler. In such rare cases if it is indeed possible we would lose some potentially possible information but for the arguably price of more convoluted and  complex code.

Reply:
I'll defer the final call to @Datadog/windows-agent so all good from my side but my 2c would be that we handle this in a more Golang idiomatic manner (maybe individual methods, a loop, and/or return of multiple error messages 🤷 ) since we only seem to display the last error message in the logs and have a different set of data collected.

Reply:
> since we only seem to display the last error message

the errors are accumulated are they not?

Reply:
I handled the errors with `log.Warnf` to be more consistent and for simplicity.